server/game/cards/01_SOR/LukeSkywalkerFaithfulFriend.ts

```diff
@@ -0,0 +1,58 @@
+import AbilityHelper from '../../AbilityHelper';
+import { Card } from '../../core/card/Card';
+import { PlayableCard } from '../../core/card/CardTypes';
+import { LeaderUnitCard } from '../../core/card/LeaderUnitCard';
+import { Aspect, RelativePlayer, Trait } from '../../core/Constants';
+import { StateWatcherRegistrar } from '../../core/stateWatcher/StateWatcherRegistrar';
+import { CardsPlayedThisPhaseWatcher } from '../../stateWatchers/CardsPlayedThisPhaseWatcher';
+
+export default class GrandMoffTarkinOversectorGovernor extends LeaderUnitCard {
+    private cardsPlayedThisPhaseWatcher: CardsPlayedThisPhaseWatcher;
+
+    protected override getImplementationId() {
+        return {
+            id: '2579145458',
+            internalName: 'luke-skywalker#faithful-friend',
+        };
+    }
+
+    protected override setupStateWatchers(registrar: StateWatcherRegistrar): void {
+        this.cardsPlayedThisPhaseWatcher = AbilityHelper.stateWatchers.cardsPlayedThisPhase(registrar, this);
+    }
+
+    protected override setupLeaderSideAbilities() {
+        this.addActionAbility({
+            title: 'Give a shield to a heroism unit you played this phase',
+            cost: [AbilityHelper.costs.abilityResourceCost(1), AbilityHelper.costs.exhaustSelf()],
+            targetResolver: {
+                controller: RelativePlayer.Self,
+                cardCondition: (card, context) => {
+                    if (!card.hasSomeAspect(Aspect.Heroism)) {
+                        return false;
+                    }
+
+                    const cardsPlayed = this.cardsPlayedThisPhaseWatcher.getCurrentValue();
+                    const playedThisTurn = cardsPlayed
+                        .filter((playedCardEntry) => playedCardEntry.playedBy === context.source.controller)
```

Comment:
Ditto to my comment on Medal Ceremony. In this case, I'm imagining something like:
this.cardsPlayedThisPhaseWatcher.getMatchingCards((playedCardEntry) => playedCardEntry.playedBy === context.source.controller);

Reply:
Oh yeah I like that! The only hiccup is that right now I'm deliberately leaving it so that the `TState` type can be literally anything, not necessarily an array of objects that can map cleanly to cards like these two. In order to be able to write a method that doesn't require the explicit map logic, we'd have to create a base `TState` like this:

```typescript
interface ICardEntryBase {
   card: Card;
}

type IWatcherStateBase<TEntry extends ICardEntryBase> = TEntry[];

class StateWatcher<TState extends IWatcherStateBase> {
// ...
```

I was originally hesitant to do something like that because I wanted to keep state open-ended but I think watchers have evolved to the point that we can be confident that this would right now with all of the cards that watch anything (I'm pretty sure they all just boil down to saving a list of cards that did something this turn).

If we make that change we can a card filter. Does that seem reasonable to you?

Reply:
Yeah, I think that approach works.